### PR TITLE
make roleplay cards more flexible; add support for instructions anywhere in card

### PR DESCRIPTION
**Pull request by toddmedema**
Instructions can now be displayed anywhere, not just grouped at the bottom.

Also lays the foundation for https://github.com/ExpeditionRPG/expedition-app/issues/160 by allowing more flexible types of elements in a roleplaying card

screen shot 2017-03-15 at 5 04 49 pm

### CODE REVIEW

File: app/QuestParser.test.tsx [hunk 1]
```diff
@@ -12,24 +12,25 @@ describe('QuestParser', () => {
...
       // Lines with nothing but variable assignment are hidden
       var result = loadRoleplayNode(cheerio.load('<roleplay><p>{{text="TEST"}}</p><p>{{text}}</p></roleplay>')('roleplay'), {scope: {}});
-      expect(mount(result.content).html()).toEqual('<span><p>TEST</p></span>');
+      expect(result.content[0].text).toEqual('<p>TEST</p>');
+      expect(result.content.length).toEqual(1);
 
       // variables with value 0 and 1 display properly
       var result = loadRoleplayNode(cheerio.load('<roleplay><p>{{a=0}}{{b=1}}</p><p>{{a}}{{b}}</p></roleplay>')('roleplay'), {scope: {}});
-      expect(mount(result.content).html()).toEqual('<span><p>01</p></span>');
+      expect(result.content[0].text).toEqual('<p>01</p>');
```
Comment (smartin015) [inline app/QuestParser.test.tsx]: I'd prefer something like:

```
expect(result.content).toEqual([{text: '<p>01</p>'}]);
```

Here and elsewhere. That way we're testing on the full body of text and not ignoring if anything somehow happens after.

File: app/QuestParser.tsx [hunk 2]
```diff
@@ -392,16 +391,12 @@ function generateIconElements(content: string): string {
...
+  let numEvents = 0;
+  let choices: Choice[] = [];
+  let choiceCount = -1;
+  let children: RoleplayElement[] = [];
 
-  var newScope = Object.assign({}, ctx.scope);
-
-  var choiceCount = -1;
-  var instructionCount = -1;
+  const newScope = Object.assign({}, ctx.scope);
```
Comment (smartin015) [inline app/QuestParser.tsx]: Spread? :D

File: app/QuestParser.tsx [hunk 3]
```diff
@@ -427,33 +422,41 @@ export function loadRoleplayNode(node: XMLElement, ctx: QuestContext): RoleplayR
...
 
-    // accumulate 'instruction' tags in tags[]
+    let type = 'text';
+    let textContent = '';
     if (tag === 'instruction') {
-      instructionCount++;
-      instructions.push({text: generateIconElements(evaluateContentOps(c.html(), newScope)), idx: instructionCount});
-      return;
+      type = 'instruction';
+      textContent = c.html();
```
Comment (smartin015) [inline app/QuestParser.tsx]: Random thought... it looks like this allows users to pass through arbitrary HTML code. Are we doing any app-side sanitizing?

File: app/QuestParser.tsx [hunk 4]
```diff
@@ -427,33 +422,41 @@ export function loadRoleplayNode(node: XMLElement, ctx: QuestContext): RoleplayR
...
-      textContent = c.toString();
-    } else {
-      throw new Error('Invalid element ' + c);
+    textContent = evaluateContentOps(textContent, newScope);
+
+    if (textContent !== '') {
+      children.push({
+        type,
+        text: generateIconElements(textContent),
+        idx: children.length,
```
Comment (smartin015) [inline app/QuestParser.tsx]: Why are we enumerating objects in a list?

File: app/reducers/QuestTypes.tsx [hunk 5]
```diff
@@ -85,7 +85,8 @@ export interface EventParameters {
   heal?: number;
 }
 
-export interface Instruction {
+export interface RoleplayElement {
+  type: string; // text, instruction
```
Comment (smartin015) [inline app/reducers/QuestTypes.tsx]: Instead: `type: 'text' | 'instruction';`

File: app/QuestParser.tsx (hunk 3 above)
Comment (toddmedema) [inline app/QuestParser.tsx]: Hmm. Should we be doing that in the app or quest creator? Pretty sure we're already doing that in the QC

File: app/reducers/QuestTypes.tsx (hunk 5 above)
Comment (toddmedema) [inline app/reducers/QuestTypes.tsx]: Wait, really? That's awesome

File: app/QuestParser.tsx (hunk 4 above)
Comment (toddmedema) [inline app/QuestParser.tsx]: Touche

File: app/QuestParser.tsx (hunk 2 above)
Comment (toddmedema) [inline app/QuestParser.tsx]: 👍

File: app/QuestParser.test.tsx (hunk 1 above)
Comment (toddmedema) [inline app/QuestParser.test.tsx]: @smartin015 updated this; all comments addressed

File: app/QuestParser.tsx (hunk 3 above)
Comment (smartin015) [inline app/QuestParser.tsx]: I'd say both. We don't know what input sources we'll end up using aside from checked-in quests, and there might be ways of adding quests later on that don't go through the sanitization (for instance, direct calls to /publish without going through the editor). 

But it also benefits us to not have latent unsanitized html code in our DB waiting for an exploit.

File: app/QuestParser.tsx (hunk 2 above)
Comment (smartin015) [inline app/QuestParser.tsx]: Doesn't look like this was spreaded...

**Comment by toddmedema** [comment]
Fixed the last spreads; mergin'

**Pull request merged**